### ISSUE TITLE
I want QoS in the UL, I realized the iperf from the UE to the network.

### ISSUE THREAD

**Issue by AEncinas19** [srsran/srsRAN_Project#353]
I created another issue due to I want QoS in the UL. At this moment, QoS in DL is not relevant.

              I want QoS in the UL, I realized the iperf from the UE to the network.

_Originally posted by @AEncinas19 in https://github.com/srsran/srsRAN_Project/issues/351#issuecomment-1838965722_

**Comment by ismagom** [comment]
What UE are you using? If it's srsUE is not supported on that side either. Any other UE please can you share mac pcap?

Thanks

**Comment by AEncinas19** [comment]
I am using COTS UE, a Samsung A13 5G. I think I cannot share a mac pcap. Should I configure something in the gNB QoS mac section?

**Comment by AEncinas19** [comment]
Could the roaming add complexity to the QoS enforcement?

**Comment by herlesupreeth** [comment]
@AEncinas19 please share the NGAP pcap when you attach the UE to 5G network. And, can you share the steps about how you performed the iperf test? I mean where did you run the iperf server and client etc?

**Comment by AEncinas19** [comment]
I have a server connected to the virbr0 (simulating that this server is in Internet) which the UPF is also connected in order to do the iperf3 tests. There, I run the iperf3 server. 

On the other hand, I run the iperf3 client in the COTS UE using the PingTools app downloaded from the Play Store.

Bellow I attatch the NGAP pcap (I've disconnected and connected the phone two times in order to show clearer all the process).

[pcap.zip](https://github.com/srsran/srsRAN_Project/files/13558416/pcap.zip)

**Comment by herlesupreeth** [comment]
Thank you the pcap. First of all, I dont think there is any rate limiting functionality implemented in gNB to restrict the rate to the one configured in 5GC, therefore whatever the rate at which UE sends that will be rate you see on iperf server. Technically UE should not be sending more than 300 kbps since that is what is sent in NAS message.

Second, I would recommend defining the Flow description in open5gs WebUI to something like below and see if it helps

`permit out from any 1-65535 to 192.168.122.5 5201`

**Comment by AEncinas19** [comment]
I changed the flow description and the behaviour it is the same. Despite in the N3 interface the traffic that matches with the flow description has a QFI=2 and different for the other traffic (which has a QFI=1), the is not any bit rate limitation on the traffic as specified in the PCC rule. 

As far as I am concerned (3GPP specifications), the gNB can guarantee and assure the GFBR and MFBR for GBR flows. Maybe, the gNB only supports the QoS signalling but not the physical implementation.

**Comment by AEncinas19** [comment]
I changed the flow description and the behaviour it is the same. Despite in the N3 interface the traffic that matches with the flow description has a QFI=2 and different for the other traffic (which has a QFI=1), the is not any bit rate limitation on the traffic as specified in the PCC rule. 

As far as I am concerned (3GPP specifications), the gNB can guarantee and assure the GFBR and MFBR for GBR flows. Maybe, the gNB only supports the QoS signalling but not the physical implementation.

**Comment by herlesupreeth** [comment]
You could alter the below parameters in `qos.yml` file for the 5QI for which you want to have bit rate limitation and see if that helps

```
      bucket_size_duration_ms: 5
      prioritized_bit_rate_kBps: 65537
```

**Comment by AEncinas19** [comment]
I changed the flow description (Uplink
permit out tcp from any 1-65535 to 192.168.122.5 5201) and the behaviour it is the same, the is not any bit rate limitation on the traffic as specified in the PCC rule.

However, with the last commit in the main branch, it is very difficult to connect the COTS UE to the network. The UE sends a Release Command after 1 minute more or less, due to unknown-local-UE-NGAP-ID. I don't know what is happening now.

As far as I am concerned (3GPP specifications), the gNB can guarantee and assure the GFBR and MFBR for GBR flows. Maybe, the gNB only supports the QoS signalling but not the physical implementation.

I will try to modify the bucket_size_duration and the prioritized_bit_rate, but with the error mentioned its very difficult to establish a PDU Session.

[ErrorIndication_NGAP.zip](https://github.com/srsran/srsRAN_Project/files/13647796/ErrorIndication_NGAP.zip)

**Comment by herlesupreeth** [comment]
Its hard to debug with the above attached pcap. I would suggest to retake the pcap for following scenario:

1. Put phone in Airplane mode
2. Start wireshark/tcpdump
3. Take out the phone from Airplane mode and try to attach
4. Stop wireshark/tcpdump

and attach here the captured trace

**Comment by AEncinas19** [comment]
Hi,

here is the captured trace.
[ErrorIndication_NGAP.zip](https://github.com/srsran/srsRAN_Project/files/13683226/ErrorIndication_NGAP.zip)

**Comment by herlesupreeth** [comment]
Please post your gNB configuration file and also gNB logs in info level

**Comment by AEncinas19** [comment]
Here is the logs and config together with the captured trace

[LogsandConfigs.zip](https://github.com/srsran/srsRAN_Project/files/13683632/LogsandConfigs.zip)

**Comment by herlesupreeth** [comment]
May I know which SIM card you are using?  If you are using Sysmocom SIM card you may want to disable 5G files on the SIM card. From your pcap I see that Security Mode Command is taking very long time to complete which is quite strange. Are you using NEA1 or NEA2 for ciphering on the AMF?

**Comment by AEncinas19** [comment]
I'm using a opencells SIM (https://open-cells.com/index.php/sim-cards/). The case is that in a previous version all worked fine, but with the new commits and trying to use QoS, my phone has connection problems.

For ciphering on the AMF I'm using:

         integrity_order : [ NIA2, NIA1, NIA0 ]
        ciphering_order : [ NEA0, NEA1, NEA2 ]

**Comment by herlesupreeth** [comment]
Then, there should not be any issue at all. Can you please post the QoS configuration you are using at gNB?

**Comment by AEncinas19** [comment]
Of course, here it is.
[qos.zip](https://github.com/srsran/srsRAN_Project/files/14138854/qos.zip)

**Comment by herlesupreeth** [comment]
Thanks. I dont see any issue with QoS configuration either. Sorry, I dont know anymore why you are facing ErrorIndication at NGAP with latest code. 

@FabianEckermann Can you please take a look at the trace here - https://github.com/srsran/srsRAN_Project/issues/353#issuecomment-1857515088 ??

**Comment by FabianEckermann** [comment]
Hi,
could you provide F1AP and NGAP PCAPs and a logfile with with the following log levels:
`log --lib_level warning --cu_level debug --f1ap_level debug --rrc_level debug --ngap_level debug`
You can check out https://docs.srsran.com/projects/project/en/latest/user_manuals/source/outputs.html for how to get logs and PCAPs from the gnb application.

**Comment by AEncinas19** [comment]
Hi,
here it is the pcaps and logs you have requested.
[Pcaps_and_logs.zip](https://github.com/srsran/srsRAN_Project/files/14164880/Pcaps_and_logs.zip)

Thank you for the support

**Comment by FabianEckermann** [comment]
Ok, I would also need mac PCAPs and logs with `log --lib_level warning --cu_level debug --f1ap_level debug --rrc_level debug --ngap_level debug --rlc_level info` (added RLC logs here)

**Comment by AEncinas19** [comment]
Here is the zip file.
[LogsandPCAPs.zip](https://github.com/srsran/srsRAN_Project/files/14167402/LogsandPCAPs.zip)

**Comment by FabianEckermann** [comment]
Looking at your logs I see a lot of lost messages, causing procedures to timeout. You should try to achieve a more stable setup first, before trying more advanced scenarios

**Comment by AEncinas19** [comment]
Well, I've updated to the last commit and improve the radio signal and now the UE connects with the 5G network perfectly.

However, the RAN does not receive any traffic descriptor in order to detect traffic and apply QoS. It receives the bit rate limitation and the 5QI, but it does not receive the traffic description. Moreover, I'm changing the prioritized_bit_rate_kBps and it does not apply any limitation to the bitrate. Is the QoS implemented in the RAN?

**Comment by AEncinas19** [comment]
![image](https://github.com/srsran/srsRAN_Project/assets/49205996/ae2f008c-548e-420a-b253-e16fe4e71447)
![image](https://github.com/srsran/srsRAN_Project/assets/49205996/4fdb5eb2-b872-4e99-9e84-3e2eaa9bc1ea)

**Comment by herlesupreeth** [comment]
> However, the RAN does not receive any traffic descriptor in order to detect traffic and apply QoS. It receives the bit rate limitation and the 5QI, but it does not receive the traffic description. 

Please send us entire NGAP, F1AP and MAC pcap for this.

> Moreover, I'm changing the prioritized_bit_rate_kBps and it does not apply any limitation to the bitrate. Is the QoS implemented in the RAN?

Please share again gNB configuration files you used. Also, please note that prioritized_bit_rate_kBps is an UL (config) parameter and technically its the UE that need to apply this limit and we dont have any rate limiter functionality in gNB

**Comment by AEncinas19** [comment]
> Please share again gNB configuration files you used. Also, please note that prioritized_bit_rate_kBps is an UL (config) parameter and technically its the UE that need to apply this limit and we dont have any rate limiter functionality in gNB

How is physical QoS implemented in srsRAN? Is signalling only implemented?

**Comment by ismagom** [comment]
Yes only signalling is implemented (QoS in the UL). QoS in the DL will be available in the 24.4 release.

**Comment by AEncinas19** [comment]
Thank you so much. All doubts are resolved.
